packages/core/src/electron-main/electron-main-application.ts

```diff
@@ -495,7 +539,22 @@ export class ElectronMainApplication {
     }
 
     protected onWindowAllClosed(event: ElectronEvent): void {
-        this.requestStop();
+        if (!this.restarting) {
+            this.requestStop();
+        }
+    }
+
+    protected restart(id: number): void {
+        this.restarting = true;
+        BrowserWindow.fromId(id).close();
+        setTimeout(async () => {
+            await this.launch({
+                secondInstance: false,
+                argv: this.processArgv.getProcessArgvWithoutBin(process.argv),
+                cwd: process.cwd()
+            });
+            this.restarting = false;
+        }, 500);
```

Comment:
What's the significance of the `setTimeout` here? Could it be extracted into some sort of one-time event listener?

Reply:
The `BrowserWindow.close` call isn't instant and therefore we have two windows open for a really short duration. This isn't a problem for the user usually, since it looks like the old window swaps with the new one. However, due to #9560, having two windows on the same spot isn't allowed. So Theia will instead move the second window to a new location. If we wait a bit, the first window will successfully close and the new window will appear at the original position.

But you're right, this can be successfully moved to the `on('closed')` event listener.